### PR TITLE
created class for denoising metrics

### PR DESCRIPTION
**Pull request by AnnaAltynova**
metrics class for denoising performance estimation. contains source - prediction metrics for a batch, fourier power spectrum for an image, and local correlation and similarity maps for denoised image - estimated noise comparison.

### CODE REVIEW

File: seismiqb/denoising_metrics.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+        image_view = image_view.reshape(straighten)
+        image_noise_view = image_noise_view.reshape(straighten)
+
+        pearson = cls._pearson_corr_2d(image_view, image_noise_view).reshape(img_shape)
+        return np.nan_to_num(pearson)
+
+    @classmethod
+    def _pearson_corr_2d(cls, x, y):
+        """Squared Pearson correlation coeffitient between corresponding rows of 2d input arrays"""
+        x_centered = x - x.mean(axis=1).reshape(-1, 1)
```
Comment (akoryagin) [inline seismiqb/denoising_metrics.py:144]: `keepdims=True` in `mean`

File: seismiqb/denoising_metrics.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        straighten = (np.dot(*image_view.shape[:2]), np.dot(*image_view.shape[2:]))
+        image_view = image_view.reshape(straighten)
+        image_noise_view = image_noise_view.reshape(straighten)
+
+        pearson = cls._pearson_corr_2d(image_view, image_noise_view).reshape(img_shape)
+        return np.nan_to_num(pearson)
+
+    @classmethod
+    def _pearson_corr_2d(cls, x, y):
+        """Squared Pearson correlation coeffitient between corresponding rows of 2d input arrays"""
```
Comment (akoryagin) [inline seismiqb/denoising_metrics.py:143]: typo: coefficient

File: seismiqb/denoising_metrics.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        x_centered = x - x.mean(axis=1).reshape(-1, 1)
+        y_centered = y - y.mean(axis=1).reshape(-1, 1)
+        corr = (x_centered * y_centered).sum(axis=1)
+        corr /= np.sqrt((x_centered**2).sum(axis=1) * (y_centered**2).sum(axis=1))
+        return corr ** 2
```
Comment (akoryagin) [inline seismiqb/denoising_metrics.py:148]: I don't think the you need to take `corr` squared

File: seismiqb/denoising_metrics.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+class DenoisingMetrics:
+    """Class for metrics and color maps assotiated with denoising performance estimation."""
+    @classmethod
+    def get_metrics(cls, images, predictions, config=None):
+        """Compute metrics related to denoising performance for a batch of images.
+
+        Parameters
+        ----------
+        images : sequence
```
Comment (akoryagin) [inline seismiqb/denoising_metrics.py:17]: Need more info here: sequence of what? Of 2d-arrays - or does it support several channels?

File: seismiqb/denoising_metrics.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+        # "same" padding along trace for 1d window or both dims for 2d
+        pad = window_size // 2
+        pad_width = [[pad, window_size - (1 + pad)], [pad * (n_dims - 1), (window_size - (1 + pad)) * (n_dims - 1)]]
+
+        image = np.pad(image, pad_width=pad_width, mode='mean')
+        image_noise = np.pad(image_noise, pad_width=pad_width, mode='mean')
+
+        # vectorization
+        window_shape=[window_size, window_size if n_dims == 2 else 1]
```
Comment (akoryagin) [inline seismiqb/denoising_metrics.py:130]: `window_shape = [window_size, window_size if n_dims == 2 else 1]`

File: seismiqb/denoising_metrics.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        image_view = np.lib.stride_tricks.sliding_window_view(image, window_shape=window_shape)
+        image_noise_view = np.lib.stride_tricks.sliding_window_view(image_noise, window_shape=window_shape)
+
+        straighten = (np.dot(*image_view.shape[:2]), np.dot(*image_view.shape[2:]))
+        image_view = image_view.reshape(straighten)
+        image_noise_view = image_noise_view.reshape(straighten)
+
+        H = np.eye(window_size**n_dims, dtype=np.float) * lamb
+        H = np.lib.stride_tricks.as_strided(H, shape=(image_view.shape[0], window_size**n_dims, window_size**n_dims),
+                                                strides=(0, 8 * window_size**n_dims, 8))
```
Comment (akoryagin) [inline seismiqb/denoising_metrics.py:199]: Fix tabulation

**Review by HollowPrincess — CHANGES_REQUESTED** [review]
Think about the code structure: which methods do we really need and in which view, what API is the most suitable for usecases.
Also, think about more appropriate and clear variable names and docstrings.

File: seismiqb/denoising_metrics.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+                   'uqi': universal_image_quality_index,
+                   'mse': mean_squared_error}
+
+        metric_names = metrics.keys() if config is None else config['metric_names']
+        if not np.all([metric in metrics for metric in metric_names]):
+            raise ValueError
+
+        config = config if config is not None else {}
+        images = images if isinstance(images, torch.Tensor) else torch.tensor(images)
+        predictions = predictions if isinstance(predictions, torch.Tensor) else torch.tensor(predictions)
```
[the remark this answers is not in the document]
Comment (AnnaAltynova) [reply seismiqb/denoising_metrics.py:47]: torchmetrics accepts torch.Tensors only.

**Pull request merged**